x-pack/filebeat/input/http_endpoint/input.go

```diff
@@ -81,28 +83,102 @@ func (e *httpEndpoint) Test(_ v2.TestContext) error {
 }
 
 func (e *httpEndpoint) Run(ctx v2.Context, publisher stateless.Publisher) error {
-	log := ctx.Logger.With("address", e.addr)
+	err := servers.serve(ctx, e, publisher)
+	if err != nil && err != http.ErrServerClosed {
+		return fmt.Errorf("unable to start server due to error: %w", err)
+	}
+	return nil
+}
 
-	mux := http.NewServeMux()
-	mux.Handle(e.config.URL, newHandler(e.config, publisher, log))
-	server := &http.Server{Addr: e.addr, TLSConfig: e.tlsConfig, Handler: mux}
-	_, cancel := ctxtool.WithFunc(ctx.Cancelation, func() { server.Close() })
-	defer cancel()
+// servers is the package-level server pool.
+var servers = pool{servers: make(map[string]*server)}
+
+// pool is a concurrence-safe pool of http servers.
+type pool struct {
+	mu sync.Mutex
+	// servers is the server pool keyed on their address/port.
+	servers map[string]*server
+}
+
+// serve runs an http server configured with the provided end-point and
+// publishing to pub. The server will run until either the context is
+// cancelled or the context of another end-point sharing the same address
+// has had its context cancelled. If an end-point is re-registered with
+// the same address and mux pattern, serve will return an error.
+func (p *pool) serve(ctx v2.Context, e *httpEndpoint, pub stateless.Publisher) error {
+	log := ctx.Logger.With("address", e.addr)
+	pattern := e.config.URL
 
 	var err error
-	if server.TLSConfig != nil {
-		log.Infof("Starting HTTPS server on %s", server.Addr)
-		// certificate is already loaded. That's why the parameters are empty
-		err = server.ListenAndServeTLS("", "")
-	} else {
-		log.Infof("Starting HTTP server on %s", server.Addr)
-		err = server.ListenAndServe()
+	p.mu.Lock()
+	s, ok := p.servers[e.addr]
+	if ok {
+		if s.pattern[pattern] {
+			err := fmt.Errorf("pattern already exists for %s: %s", e.addr, pattern)
```

Comment:
I think it will be helpful to log the IDs of the inputs that caused the conflicts. Under Agent these IDs would fully identify the associated integrations and this would help facilitate quick resolution of the conflict.

Reply:
Nice. I'll add that. This is `ctx.Agent.ID` correct? Maybe the whole `beat.Info` for the agent?

Reply:
I'm talking about the input context ID. It will have values like `http_endpoint-zoom.foo-ac0dab2c-e54f-4964-9104-e3ca9104516f`.

https://github.com/elastic/beats/blob/ba3bce42590dbf722061d7d92bfdb0bd903e9014/filebeat/input/v2/input.go#L85-L86.

Reply:
Done